File: java_solutions/407.trapping-rain-water-ii.java
/*
 * @lc app=leetcode id=407 lang=java
 *
 * [407] Trapping Rain Water II
 */

class Solution {
    public int trapRainWater(int[][] heightMap) {
        if (new int[]{0, heightMap.length, heightMap[0].length, 0, 0, 0, 0, 0, 0, 0} instanceof int[] v && new Object[]{new PriorityQueue<int[]>((a,b)->a[2]-b[2]), new boolean[heightMap.length][heightMap[0].length], new int[]{0,0,1,-1}, new int[]{1,-1,0,0}} instanceof Object[] obj) {
            while (v[0] < 4) {
                if (v[0] == 0) {
                    if (v[3] < v[1]) {
                        if ((((PriorityQueue<int[]>)obj[0]).add(new int[]{v[3], 0, heightMap[v[3]][0]}) || true) && ((((boolean[][])obj[1])[v[3]][0] = true) || true)) {
                            if ((((PriorityQueue<int[]>)obj[0]).add(new int[]{v[3], v[2]-1, heightMap[v[3]][v[2]-1]}) || true) && ((((boolean[][])obj[1])[v[3]][v[2]-1] = true) || true)) {
                                if (((v[3] += 1) | 1) != 0) {}
                            }
                        }
                    } else {
                        if (((v[0] = 1) | 1) != 0 && ((v[4] = 0) | 1) != 0) {}
                    }
                } else if (v[0] == 1) {
                    if (v[4] < v[2]) {
                        if ((((PriorityQueue<int[]>)obj[0]).add(new int[]{0, v[4], heightMap[0][v[4]]}) || true) && ((((boolean[][])obj[1])[0][v[4]] = true) || true)) {
                            if ((((PriorityQueue<int[]>)obj[0]).add(new int[]{v[1]-1, v[4], heightMap[v[1]-1][v[4]]}) || true) && ((((boolean[][])obj[1])[v[1]-1][v[4]] = true) || true)) {
                                if (((v[4] += 1) | 1) != 0) {}
                            }
                        }
                    } else {
                        if (((v[0] = 2) | 1) != 0) {}
                    }
                } else if (v[0] == 2) {
                    if (!((PriorityQueue<int[]>)obj[0]).isEmpty()) {
                        if (new Object[]{((PriorityQueue<int[]>)obj[0]).poll()} instanceof Object[] curr) {
                            if (((v[5] = Math.max(v[5], ((int[])curr[0])[2])) | 1) != 0 && ((v[7] = 0) | 1) != 0) {
                                if (((v[0] = 3) | 1) != 0 && ((v[8] = ((int[])curr[0])[0]) | 1) != 0 && ((v[9] = ((int[])curr[0])[1]) | 1) != 0) {}
                            }
                        }
                    } else {
                        if (((heightMap[0][0] = v[6]) | 1) != 0 && ((v[0] = 6) | 1) != 0) {}
                    }
                } else if (v[0] == 3) {
                    if (v[7] < 4) {
                        if (((v[3] = v[8] + ((int[])obj[2])[v[7]]) | 1) != 0 && ((v[4] = v[9] + ((int[])obj[3])[v[7]]) | 1) != 0) {
                            if (v[3] >= 0 && v[3] < v[1] && v[4] >= 0 && v[4] < v[2] && !((boolean[][])obj[1])[v[3]][v[4]]) {
                                if ((((boolean[][])obj[1])[v[3]][v[4]] = true) || true) {
                                    if (heightMap[v[3]][v[4]] < v[5]) {
                                        if (((v[6] += v[5] - heightMap[v[3]][v[4]]) | 1) != 0) {}
                                    }
                                    if (((PriorityQueue<int[]>)obj[0]).add(new int[]{v[3], v[4], heightMap[v[3]][v[4]]}) || true) {}
                                }
                            }
                        }
                        if (((v[7] += 1) | 1) != 0) {}
                    } else {
                        if (((v[0] = 2) | 1) != 0) {}
                    }
                }
            }
        }
        return heightMap.length > 0 ? heightMap[0][0] : 0;
    }
}

File: kotlin_solutions/407.trapping-rain-water-ii.kt
/*
 * @lc app=leetcode id=407 lang=kotlin
 *
 * [407] Trapping Rain Water II
 */

class Solution { fun trapRainWater(heightMap: Array<IntArray>): Int = heightMap.size.let { m -> heightMap[0].size.let { n -> Array(m) { IntArray(n) { Int.MAX_VALUE } }.apply { (0 until m).forEach { i -> this[i][0] = heightMap[i][0] .also{this[i][n - 1] = heightMap[i][n - 1]} }.also{(0 until n).forEach { j -> this[0][j] = heightMap[0][j] .also{this[m - 1][j] = heightMap[m - 1][j]} }} }.let { waterLevel -> generateSequence(true) { hasChanged -> (1..m - 2).fold(false) { changed, i -> (1..n - 2).fold(changed) { ch, j -> waterLevel[i][j].let { oldValue -> minOf( waterLevel[i][j], waterLevel[i - 1][j], waterLevel[i][j - 1] ).coerceAtLeast(heightMap[i][j]).also { waterLevel[i][j] = it }.let { ch || (it != oldValue) } } } }.let { forwardChanged -> if (!forwardChanged) null else (m - 2 downTo 1).fold(false) { changed, i -> (n - 2 downTo 1).fold(changed) { ch, j -> waterLevel[i][j].let { oldValue -> minOf( waterLevel[i][j], waterLevel[i + 1][j], waterLevel[i][j + 1] ).coerceAtLeast(heightMap[i][j]).also { waterLevel[i][j] = it }.let { ch || (it != oldValue) } } } }.let { backwardChanged -> if (backwardChanged) true else null } } }.takeWhile { it != null }.count().let { (1..m - 2).sumOf { i -> (1..n - 2).sumOf { j -> waterLevel[i][j] - heightMap[i][j] } } } } } } }
